Reorder match optimizer

Leo-Besancon opened the pull request:
This PR closes https://github.com/0xMiden/air-script/issues/431

Task list:
- [x] Implement UnrollingThirdPass to handle `If` nodes
- [x] Add E2E test mixing selectors to combine and list comprehension
- [x] Check the produced graph and ensure the match optimizer correctly handled this case.
- [x] Clean and refactor codeto avoid redundancy between UnrollingFirstPass and UnrollingThirdPass.

Note: This branch is currently based on top of PR https://github.com/0xMiden/air-script/pull/434, it will need to be rebased on next when it is merged.

Leo-Besancon committed:
refactor: split the two Unrolling passes, add basic helper functions

Leo-Besancon committed:
refactor: Isolate MatchOptimizer in separate struct

Leo-Besancon committed:
refactor(unrolling): add helper functions for For nodes handling

Leo-Besancon committed:
chore: Update comment for Unrolling pass

Leo-Besancon committed:
chore: add comment for visit_*_bis() dispatch

Leo-Besancon committed:
chore: add backticks around node types in comments

Leo-Besancon committed:
chore: remove empty impl blocks for two structures

Leo-Besancon committed:
refactor: Instantiate one MatchOptimizer per `If` node

Leo-Besancon committed:
chore: add newlines

Leo-Besancon committed:
chore: Improve various comments based on review

Leo-Besancon committed:
feat: Implement Unrolling 3rd pass for MatchOptimizer

Leo-Besancon committed:
tests: Add selector combine test with list comprehensions

Leo-Besancon committed:
chore: Update CHANGELOG.md

Leo-Besancon committed:
chore: update CHANGELOG.md

adr1anh commented on air-script/tests/selectors/selectors_combine_with_list_comprehensions.air:
```diff
@@ -0,0 +1,59 @@
+    enf c.first = 0;
+}
+
+# Note: we want the following constraints values to be equivalent:
```
I don't fully understand what this comment means at first glance. What are we testing/expecting here?

adr1anh commented on mir/src/passes/unrolling/unrolling_third_pass.rs:
```diff
@@ -0,0 +1,468 @@
+        _graph: &mut Graph,
+        boundary: Link<Op>,
+    ) -> Result<Option<Link<Op>>, CompileError> {
+        let mut updated_boundary = None;
```
Nit(also applicable to the other visit functions): this variable can be omitted by returning the result in the `if` branch.

adr1anh commented on mir/src/passes/unrolling/unrolling_third_pass.rs:
```diff
@@ -0,0 +1,468 @@
+            let iterator_nodes = iterator_vector.children().borrow().deref().clone();
+
+            let mut acc_node = initial_value;
+            for iterator_node in iterator_nodes {
```
Nit: we can replace the loop with `fold`

adr1anh commented on mir/src/passes/unrolling/unrolling_third_pass.rs:
```diff
@@ -0,0 +1,468 @@
+        _graph: &mut Graph,
+        vector: Link<Op>,
+    ) -> Result<Option<Link<Op>>, CompileError> {
+        let mut updated_vector = None;
```
Return the `Option` from the `if` branch

adr1anh commented on mir/src/passes/unrolling/unrolling_third_pass.rs:
```diff
@@ -0,0 +1,468 @@
+            if let Op::Vector(vec) = expr.borrow().deref() {
+                let ops = vec.children().borrow().deref().clone();
+                let mut new_vec = vec![];
+                for op in ops.iter() {
```
Nit: use `Vec::collect`

adr1anh commented on mir/src/passes/unrolling/unrolling_third_pass.rs:
```diff
@@ -0,0 +1,468 @@
+        value: Link<Op>,
+    ) -> Result<Option<Link<Op>>, CompileError> {
+        // safe to unwrap because we just dispatched on it
+        let mut updated_value = None;
```
Nit: return result from `match`

adr1anh commented on mir/src/passes/unrolling/unrolling_third_pass.rs:
```diff
@@ -0,0 +1,468 @@
+    parent: Link<Op>,
+    span: SourceSpan,
+) -> Result<Option<Link<Op>>, CompileError> {
+    let mut updated_binary_op = None;
```
Return result from `if` branches

adr1anh reviewed (commented):
Mostly left some nits, but otherwise looks good.

Leo-Besancon committed:
chore: clarify selector_combine comment

Leo-Besancon committed:
refactor: Improve UnrollingThirdPass visit_* functions based on review

Leo-Besancon commented:
> Mostly left some nits, but otherwise looks good.

Thank you, I've refactored the visit functions and improved the test comment.

Leo-Besancon committed:
refactor: Improve first pass based on #444 review comments

Leo-Besancon commented:
Edit: fixed several issues in the latest commits.
- The E2E test was too simple, and the list comprehensions were folded into constant vectors during AST's constant propagation, so it did not effectively test the reorder.
- The lowering of Let statements into MIR was incorrect
- The handling of ListComprehensions in AST's ConstantPropagation also had a bug
- I removed the handling of If nodes in the First pass of Unrolling

With a more suitable test, it seems to fail for now, I'll investigate and fix it!

Leo-Besancon committed:
fix: Fix AST's ConstantPropagation that caused conflict in local variables

Leo-Besancon committed:
fix: Do not unroll If nodes in Unrolling First Pass

Leo-Besancon committed:
Fix: Fix lowering of Let statements into MIR

Leo-Besancon committed:
tests: Improve selectors combine test to have non-constant List comprehension

Leo-Besancon committed:
feat: Add a pass for common subexpression elimination in the constraint graph (#419)

Leo-Besancon committed:
Merge branch 'next' into refactor_split_unrolling_pass

Leo-Besancon committed:
chore(match_optimizer): improve comments following review

Leo-Besancon committed:
chore: add example for ConstraintEvaluationMap

Leo-Besancon committed:
Merge branch 'refactor_split_unrolling_pass_upstream' into reorder_match_optimizer_upstream

Leo-Besancon committed:
Merge branch 'next' into reorder_match_optimizer_upstream

Leo-Besancon committed:
tests: update E2E test after including CSE in merge

Leo-Besancon committed:
Merge branch 'next' into reorder_match_optimizer_upstream

bobbinth commented on mir/src/passes/unrolling/unrolling_third_pass.rs:
```diff
@@ -0,0 +1,441 @@
+        }
+        Vector::create(vec, span)
+    }
+}
```
This function seems to be identical to the function with the same name in the `unrolling_first_past.rs` file. Is there a reason not to have it just in the `src/passes/unrolling/mod.rs` file and then used in both the first and the third pass files?

This applies to some other functions in this file as well.

Leo-Besancon replied:
This is no longer needed now that I reuse the same `visit_*_bis` functions from `unrolling_first_pass.rs` in `unrolling_third_pass.rs` below.

bobbinth commented on mir/src/passes/unrolling/unrolling_third_pass.rs:
```diff
@@ -0,0 +1,441 @@
+        let rhs = add_ref.rhs.clone();
+
+        unroll_binary_op(lhs, rhs, add.clone(), add_ref.span())
+    }
```
Why does this method need to take `self` or `_graph` as parameters? Seems like they are not really needed here, and then this could be just a pure function. This pure function can then be shared between this and first passes.

This applies to some other methods in this file too.

Leo-Besancon replied:
This was to replicate the signature for `visit_*(...)` methods in visitor.rs, but I agree it's not needed. I've removed these two arguments, marked them as public in unrolling_first_pass, and imported them in unrolling_third_pass in e1098d04.

I've also removed the unnecessary `to_link_and()` helper function as it can now be achieved with `map_or`.

bobbinth commented on mir/src/passes/unrolling/unrolling_third_pass.rs:
```diff
@@ -0,0 +1,441 @@
+    rhs: Link<Op>,
+    parent: Link<Op>,
+    span: SourceSpan,
+) -> Result<Option<Link<Op>>, CompileError> {
```
Could we add doc comments explaining what is supposed to happen in this function? My understanding is that this is specifically for unrolling binary operations where operands are vectors, right? If so, how can this be expressed in AirScript (I may be misremembering something, but I didn't think we could apply binary operations to vectors)?

Also, if either `lhs` or `rhs` is not a vector, the function would just return `None` - but shouldn't it be an error? Or is the situation where this function is called with `lhs` being a vector and `rhs` not being a vector is impossible? (if so, would be good to explain the the doc comments why this is the case).

Leo-Besancon replied:
Thank you for pointing these out!

You are correct, we currently cannot directly add two vectors in the frontend (instead of `v + u`, we would do something like `[v + u for (v, u) in (v, u)]`).
During the design phase for the MIR pipeline, I think we talked about the possibility of supporting in the MIR such operations, but I don't remember if it was clear whever we wanted it or not. It was a long time ago and maybe we'd rather simplify this and remove this possibility.

Note that when testing, I've also encountered one case where we rely on this: https://github.com/0xMiden/air-script/blob/next/mir/src/tests/list_comprehension.rs#L31
`enf expected = result;` with two vectors will enforce that the vector `Sub(expected - result)` equals to [0, 0, 0, 0].

Our choices are:
- Either we remove this support, make sure we catch this during semantic analysis, and rewrite the test as:
`enf expected = result for (expected, result) in (expected, result);` I'm leaning towards this one)
- Or we allow them and document them accordingly (note that `Mul` and `Exp` between two vectors are also handled element by element in the MIR)

Leo-Besancon replied:
> Also, if either lhs or rhs is not a vector, the function would just return None - but shouldn't it be an error? Or is the situation where this function is called with lhs being a vector and rhs not being a vector is impossible? (if so, would be good to explain the the doc comments why this is the case).

The `Ok(None)` case is to handle the case where we have both lhs and rhs scalars: `Add(1, 2)` is already unrolled so we have nothing to do.
		
If one of them is a vector and not the other, I agree it should be unreachable and it's probably a good idea to catch it there.

Although:
- It depends on whever we keep these functions or not, I'll wait for you input on this
- It may be better handled with direct typing checks (depending on the Typing PR)

bobbinth replied:
> we remove this support, make sure we catch this during semantic analysis, and rewrite the test as:
>   `enf expected = result for (expected, result) in (expected, result);` I'm leaning towards this one)

Yes, let's do this - though, I'd do this in a follow-up PR.

Leo-Besancon replied:
Ok, here is the corresponding issue: https://github.com/0xMiden/air-script/issues/458

bobbinth commented on mir/src/passes/unrolling/unrolling_third_pass.rs:
```diff
@@ -0,0 +1,441 @@
+        _graph: &mut Graph,
+        _matrix: Link<Op>,
+    ) -> Result<Option<Link<Op>>, CompileError> {
+        Ok(None) // Matrix are already unrolled, we have nothing to do
```
If matrixes are already unrolled, should this be `unreachable`? And do we need this function at all?

Leo-Besancon replied:
In most cases, matrices are constants and already accessed / unrolled during the AST's constant propagation. However, there are some cases (e.g. https://github.com/0xMiden/air-script/blob/next/air/src/tests/variables.rs#L131, but also later when we'll have computed indices on matrices) where we have Matrix nodes in the graph at that point.

The unrolling pass:
- Keeps scalar as is (or compute them, for instance with `Fold` nodes)
- Converts Vector-shaped operations (that are not yet Vector) into Vector nodes. This applies to e.g. `If`, `For`, or currently an `Add` on two vectors.
- Converts Matrix-shaped operations (that are not yet Matrix) into Matrix nodes. This applies to e.g. Values that are constant matrices.

So what I meant by the comment is that `Matrix` nodes are already in their primitive forms. Do you see how we can clarify this?

In aa7ca120, I've removed the function and just return Ok(None) where it is called for simplicity.

bobbinth reviewed (commented):
Looks good! Thank you. I reviewed all code except for actually checking if the codegen of the new test is correct and left some comments inline. The comments are about reducing code duplication and also simplifying the code and improving its readability.

Leo-Besancon committed:
refactor: Improved error handling with using expect / unwrap_or_else

Leo-Besancon committed:
refactor: remove code duplication between unrolling 1st and 3rd pass

Al-Kindi-0 commented on mir/src/passes/unrolling/unrolling_third_pass.rs:
```diff
@@ -0,0 +1,441 @@
+    /// the match arms, and combining them to optimize the resulting vector of constraints if
+    /// possible. We handle bus related constraints separately, as they cannot be combined with
+    /// main trace constraints.
+    fn visit_if_bis(
```
Could we add a few words about what this method returns ?

Al-Kindi-0 commented on mir/src/passes/unrolling/unrolling_third_pass.rs:
```diff
@@ -0,0 +1,441 @@
+        let children = vector_ref.elements.borrow().deref().clone();
+        let size = vector_ref.size;
+
+        if size == 1 {
```
A comment here would be helpful

Al-Kindi-0 reviewed (commented):
Did an initial pass and it looks good.
Would be good to get some more clarity on the 'unroll_binary_op' function in relation to the point raised by @bobbinth

Leo-Besancon committed:
refactor: remove visit_matrix_bis function

Leo-Besancon committed:
chore: Improve comments following review

bobbinth commented on mir/src/passes/unrolling/match_optimizer.rs:
```diff
@@ -120,12 +120,8 @@ impl<'a> MatchOptimizer<'a> {
-                            "Error: Reference to an evaluation index not found in constraints_evaluation_indices"
-                        );
-                    };
+                    let constraints_vec = self.constraints_evaluation_indices.get_mut(&index).expect("Error: An existing evaluation index was not found in constraints_evaluation_indices");
```
nit: formatting here looks a bit off.

bobbinth commented on mir/src/passes/unrolling/unrolling_third_pass.rs:
```diff
@@ -0,0 +1,148 @@
+    }
+}
+
+// For the first pass of Unrolling, we use a tweaked version of the Visitor trait,
+// each visit_*_bis function returns an Option<Link<Op>> instead of Result<(), CompileError>,
+// to mutate the nodes (e.g. modifying a Operation<Vectors> to Vector<Operations>)
+impl UnrollingThirdPass<'_> {
```
The comment refers to the "first pass" - is this correct?

Leo-Besancon replied:
I've updated the comment to refer more to what the 3rd pass does.

bobbinth commented on mir/src/passes/unrolling/unrolling_third_pass.rs:
```diff
@@ -0,0 +1,148 @@
+        &mut self.work_stack
+    }
+
+    // We visit all boundary constraints and all integrity constraints
+    // No need to visit the functions or evaluators, as they should have been inlined before this
+    // pass
+    fn root_nodes_to_visit(&self, graph: &Graph) -> Vec<Link<Node>> {
```
Should we add some `assert!` statements here to make sure that functions and evaluators have already been inlined? Or is this already handled in the `visit_node()` below?

Leo-Besancon replied:
Yes this is already done in the visit_node() function.

Instead of relying on panics inside a pass whenever its assumptions are broken, maybe we could have a way to check them earlier. Here, it could look like:
- a function `Inlining::assert_pass_results()` that ensures all functions and evaluators have been inlined,
- or a function `Unrolling::assert_pre_pass_assumptions()` that does the same.

I'm not sure if it's really useful (we don't have too much to check for now I think), but I could make an issue for this if you want.

bobbinth commented on mir/src/passes/unrolling/unrolling_first_pass.rs:
```diff
@@ -660,21 +386,155 @@ impl Visitor for UnrollingFirstPass<'_> {
-        f(graph, op)
-    } else {
-        Ok(None)
+pub fn visit_enf_bis(enf: Link<Op>) -> Result<Option<Link<Op>>, CompileError> {
```
Since this (and other similar functions) are used both here and in the third pass, it probably makes sense to move them to the parent module (i.e., to `src/passes/unrolling/mod.rs`.

bobbinth commented on mir/src/passes/unrolling/unrolling_third_pass.rs:
```diff
@@ -0,0 +1,148 @@
+                    node
+                );
+            },
+        };
```
A few nits here (could be addresses in a follow-up):

- I would change the type of `updated_op` to just `Option<Link<Op>>`. This does mean adding a `?` operators at the end of most match arms, but I think it looks a bit cleaner.
- What is the difference here between arms that return `None` vs. the ones that are unreachable? For example, why is `Node::Matrix` gets converted to `None`, even though the comment says that matrixes are already unrolled. From what I can tell, making `Node::Matrix` into `unreachable` doesn't cause any issues (or at least all the tests are passing).
- Instead of spelling out everything that is unreachable, should we just use a wildcard match?

Summarizing the above, this could look something like:
```rust
let updated_op: Option<Link<Op>> = match node.borrow().deref() {
    Node::Enf(e) => e.to_link().map_or(Ok(None), visit_enf_bis)?,
    Node::Boundary(b) => b.to_link().map_or(Ok(None), visit_boundary_bis)?,
    Node::Add(a) => a.to_link().map_or(Ok(None), visit_add_bis)?,
    Node::Sub(s) => s.to_link().map_or(Ok(None), visit_sub_bis)?,
    Node::Mul(m) => m.to_link().map_or(Ok(None), visit_mul_bis)?,
    Node::Exp(e) => e.to_link().map_or(Ok(None), visit_exp_bis)?,
    Node::Fold(f) => f.to_link().map_or(Ok(None), visit_fold_bis)?,
    Node::Vector(v) => v.to_link().map_or(Ok(None), visit_vector_bis)?,
    Node::Value(v) => v.to_link().map_or(Ok(None), visit_value_bis)?,
    Node::If(i) => i.to_link().map_or(Ok(None), |el| self.visit_if_bis(el))?,
    Node::BusOp(_b) => None,
    Node::Accessor(_a) => None,
    _ => {
        unreachable!(
            "Unexpected node during Unrolling: Function, Evaluators, Calls, For nodes and Parameters should have been inlined before this pass. Found: {:?}",
            node
        );
    },
};
```

And all the tests are still passing. Though, it is not really clear to me why `Node::BusOp` and `Node::Accessor` as unreachable causes errors.

Leo-Besancon replied:
I've fixed your first and third points in the latest commit.
For your second point:
- `Ok(None)` would mean that we can have this node type in the graph, and we do nothing if we find them (they are already in the state we want them in).
- `unreachable` would mean that we do not expect to encounter this node type in the graph.

The comment `matrixes are already unrolled` refers to the fact that if we encounter them, then we should do nothing (similarly we do nothing with vectors of size greater than 1).

For instance, if we replace the `Ok(None)` by an `unreachable` for matrices in the `UnrollingFirstPass`, this following test fails:
https://github.com/0xMiden/air-script/blob/reorder_match_optimizer_upstream/air/src/tests/variables.rs#L46

It does not fail during the third pass because the second pass already handles them, but I think we could still encounter them.

bobbinth reviewed (approved):
Looks good! Thank you! I left some comments inline - but the ones that require non-trivial amount of work, we can address in a follow-up.

Leo-Besancon committed:
chore: update comment and formating

Leo-Besancon committed:
refactor(mir): clean up visit_node matches

Leo-Besancon committed:
refactor: move visit_*_bis shared by 1st and 3rd passes into common module

Leo-Besancon committed:
Merge branch 'next' into reorder_match_optimizer_upstream

bobbinth reviewed (approved):
Looks good! Thank you! We can probably refactor things a bit more, but let's leave it for follow-up PRs.
